File: rust/crates/babylon-tick/content/scenarios/metabolism_ceiling_suppression_conformance.py
"""Conformance vector for `metabolism/biocapacity-update`'s
regeneration-suppression-AT-THE-CEILING boundary (F3 fix round, adversarial
review of PR #501: mutating `metabolism.bsl`'s `(>= current max-cap)` guard
to `(> current max-cap)` left all 16 prior tests green, because
`zero-floor-county`'s discriminating case floors BOTH branches' results to
the identical `0.0` -- the mutation was hidden by the downstream `max(0.0,
...)` clamp, not caught by it).

This script is the PROVENANCE of every number pinned in
``rust/crates/babylon-tick/tests/metabolism_ceiling_suppression_conformance.rs``.

The frozen formula's own guard, verbatim (`metabolic_rift.py:42-44`):

.. code-block:: python

    regeneration = regeneration_rate * max_biocapacity
    if current_biocapacity >= max_biocapacity:
        regeneration = 0.0

At ``current_biocapacity == max_biocapacity`` EXACTLY, this suppresses
regeneration (the ``>=`` fires). A rule using ``>`` instead would WRONGLY
let regeneration fire at this exact boundary. To make that difference
survive to the FINAL clamped output (not get floored away like
``zero-floor-county``'s vector), this territory is chosen so the CORRECT
branch (`regeneration` suppressed) lands JUST BELOW zero (and floors to
`0.0`) while the WRONG branch (`regeneration` firing) lands comfortably
ABOVE zero (and survives the floor untouched) -- ``entropy_factor`` near its
declared floor (`1.005`, so the ecological cost barely exceeds the stock
itself) and a boosted ``regeneration_rate`` (`0.5`) make the gap between
the two branches large and unambiguous.

Run it from the repository root, single process::

    PYTHONPATH="$PWD/src" uv run python \\
        rust/crates/babylon-tick/content/scenarios/metabolism_ceiling_suppression_conformance.py
"""

from __future__ import annotations

from babylon.config.defines import MetabolismDefines
from babylon.engine.context import TickContext
from babylon.engine.services import ServiceContainer
from babylon.engine.systems.metabolism import MetabolismSystem
from babylon.models.enums import NodeType
from babylon.topology.graph import BabylonGraph

SUBJECT_ID = "at-ceiling-county"
SEED = {
    "biocapacity": 10.0,
    "max_biocapacity": 10.0,
    "extraction_intensity": 1.0,
    # Boosted -- legal per Territory.regeneration_rate's own
    # Field(ge=0.0, le=1.0).
    "regeneration_rate": 0.5,
}
#: Barely above the declared floor -- legal, ordinary construction.
ENTROPY_FACTOR = 1.005
#: Legal maximum (MetabolismDefines.hysteresis_rate: Field(ge=0.001, le=0.01)).
HYSTERESIS_RATE = 0.01


def build_graph() -> BabylonGraph:
    """Build the one-territory world, seeded EXACTLY at its own ceiling."""
    graph = BabylonGraph()
    graph.add_node(SUBJECT_ID, NodeType.TERRITORY, **SEED)
    return graph


def main() -> None:
    """Run one tick of the frozen MetabolismSystem at the current==max_biocapacity boundary."""
    services = ServiceContainer.create()
    try:
        object.__setattr__(
            services.defines,
            "metabolism",
            MetabolismDefines(entropy_factor=ENTROPY_FACTOR, hysteresis_rate=HYSTERESIS_RATE),
        )

        graph = build_graph()
        MetabolismSystem().step(graph, services, TickContext(tick=1))

        node = graph.get_node(SUBJECT_ID)
        if node is None:
            raise SystemExit(f"node {SUBJECT_ID} vanished during the tick")
        a = node.attributes
        print(f"entropy_factor = {ENTROPY_FACTOR!r}, hysteresis_rate = {HYSTERESIS_RATE!r}")
        print("post-tick state:")
        print(
            f"  {SUBJECT_ID:<20} biocapacity={a['biocapacity']!r} "
            f"max_biocapacity={a['max_biocapacity']!r}"
        )
    finally:
        services.database.close()


if __name__ == "__main__":
    main()

File: rust/crates/babylon-tick/tests/metabolism_ceiling_suppression_conformance.rs
//! The AT-THE-CEILING conformance suite for `metabolism/biocapacity-update`
//! — proves the `(>= current max-cap)` regeneration-suppression guard fires
//! at the EXACT boundary `current == max-cap`, not just "close to it" (F3
//! fix round, adversarial review of PR #501).
//!
//! # Why this suite exists
//!
//! `metabolism_conformance.rs::zero_extraction_leaves_hysteresis_completely_inert`
//! and its sibling `heavy_extraction_at_the_ceiling_floors_biocapacity_at_exactly_zero`
//! both use `zero-floor-county` (`biocapacity == max_biocapacity == 100`,
//! `entropy_factor` at the production default `1.2`), whose discriminating
//! case floors BOTH the correct branch's result (`-20`) AND the wrong
//! branch's result (`-18`, if regeneration wrongly fired) to the identical
//! `0.0` — the `max(0.0, ...)` clamp downstream masks the guard's boundary
//! entirely. This scenario is built specifically so the two branches land
//! on DIFFERENT sides of the floor.
//!
//! # Provenance
//!
//! ```text
//! PYTHONPATH="$PWD/src" uv run python \
//!     rust/crates/babylon-tick/content/scenarios/metabolism_ceiling_suppression_conformance.py
//! ```
//!
//! Its output on 2026-08-11, verbatim:
//!
//! ```text
//! entropy_factor = 1.005, hysteresis_rate = 0.01
//! post-tick state:
//!   at-ceiling-county    biocapacity=0.0 max_biocapacity=9.9
//! ```

use babylon_graph::memory::MemoryGraph;
use babylon_graph::substrate::{GraphSubstrate, NodeId};
use babylon_tick::{run_once, run_once_into};

const SCENARIO: &str =
    include_str!("../content/scenarios/metabolism-ceiling-suppression-conformance.bscn");
const RULE: &str = include_str!("../content/rules/metabolism.bsl");

fn run() -> MemoryGraph {
    let mut graph = MemoryGraph::new();
    let mut sink = babylon_bsl::structural_verbs::CollectingSink::default();
    run_once_into(SCENARIO, RULE, &mut graph, &mut sink).expect("the Metabolism pack must run");
    graph
}

fn attribute(graph: &MemoryGraph, id: u64, field: &str) -> f64 {
    graph
        .node_attribute(NodeId(id), field)
        .unwrap_or_else(|e| panic!("node {id} field {field}: {}", e.message))
}

/// At `current == max_cap` exactly, regeneration is suppressed
/// (`regeneration = 0.0`), so `current + delta = 10 - 10.05 = -0.05`,
/// which floors to exactly `0.0`. Mutation-verified (by hand during
/// authoring, reverted before commit): flipping `metabolism.bsl`'s
/// `regeneration` binding's guard from `(>= current max-cap)` to
/// `(> current max-cap)` — so regeneration WRONGLY fires at this exact
/// boundary — flips this test's result to `4.949999999999999` (the guard
/// no longer suppresses the `regeneration_rate * max_cap = 5` term, which
/// survives the floor untouched).
#[test]
fn regeneration_is_suppressed_exactly_at_the_ceiling_boundary() {
    let graph = run();
    assert_eq!(attribute(&graph, 0, "territory/biocapacity"), 0.0);
}

/// The hysteresis ratchet is unaffected by this guard — `max_biocapacity`
/// ends at `9.9` regardless of which branch of the regeneration guard
/// fires, since `damage` never depends on `regeneration`.
#[test]
fn the_ceiling_still_ratchets_regardless_of_the_regeneration_guard() {
    let graph = run();
    assert_eq!(attribute(&graph, 0, "territory/max-biocapacity"), 9.9);
}

/// Byte-determinism.
#[test]
fn the_ceiling_suppression_scenario_tick_is_deterministic() {
    let a = run_once(SCENARIO, RULE).expect("first run");
    let b = run_once(SCENARIO, RULE).expect("second run");
    assert_eq!(a.after, b.after);
    assert_eq!(a.fired, 1);
}
